floodviz/static/js/map.js

```diff
@@ -1,187 +1,331 @@
+							return '';
+						}
+					})
+					.attr("class", classname)
+					.each(function (d) {
+						if (property_for_id && d.properties[property_for_id]) {
+							if (FV.hydrograph_display_ids.indexOf(d.properties.id) !== -1) {
+								self.site_add_accent(d.properties.id);
+							}
+						}
+					});
+				return (group);
+			};
+			/**
+			 * Add paths to the map
+			 * @param data The geojson to be added to the svg
+			 * @param classname The class to be given to each element for use in CSS
+			 */
+			var add_paths = function (data, classname) {
+				var group = svg.append("g");
+				group.selectAll("path")
+					.data(data.features)
+					.enter()
+					.append("path")
+					.attr("d", path)
+					.attr("class", classname);
+				return group;
+			};
+
+			/**
+			 * Starts a click-and-drag box to select gages
+			 * @param point Location of the mouse on the svg
+			 */
+			var select_box_start = function (point) {
+				if (d3.event.defaultPrevented) return;
+				svg.append('rect')
+					.attr('x', point[0])
+					.attr('y', point[1])
+					.attr('height', 0)
+					.attr('width', 0)
+					.attr('class', 'select-box')
+					.attr('id', 'map-select-box');
+
+				state.box = {
+					x: point[0],
+					y: point[1],
+					height: 0,
+					width: 0
+				};
+			};
+
+			/**
+			 * Change the size and shape of the selection box based on new mouse location
+			 * @param point Location of the mouse on the svg
+			 */
+			var select_box_drag = function (point) {
+				var box = d3.select('#map-select-box');
+				if (!box.empty()) {
+					var d = {
+						x: parseInt(box.attr("x")),
+						y: parseInt(box.attr("y")),
+						width: parseInt(box.attr("width")),
+						height: parseInt(box.attr("height"))
+					};
+					var move = {
+						x: point[0] - d.x,
+						y: point[1] - d.y
+					};
+
+					if (move.x < 1 || (move.x * 2 < d.width)) {
+						d.x = point[0];
+						d.width -= move.x;
 					}
-					else{
-						return '';
+					else {
+						d.width = move.x;
 					}
-				})
-				.attr("class", classname)
-				.each(function (d) {
-					if (property_for_id && d.properties[property_for_id]) {
-						if (FV.hydrograph_display_ids.indexOf(d.properties.id) !== -1) {
-							self.site_add_accent(d.properties.id);
-						}
+
+					if (move.y < 1 || (move.y * 2 < d.height)) {
+						d.y = point[1];
+						d.height -= move.y;
+					}
+					else {
+						d.height = move.y;
 					}
+
+					box.attr('x', d.x)
+						.attr('y', d.y)
+						.attr('width', d.width)
+						.attr('height', d.height);
+
+					state.box = {
+						x: d.x,
+						y: d.y,
+						height: d.height,
+						width: d.width
+					};
+				}
+			};
+
+			/**
+			 * End the selection box, get rid of the box in state and on the svg,
+			 * update the hydrograph and map if necessary
+			 */
+			var select_box_end = function () {
+				// Check if the box has a reasonable area to make sure it isn't a click by mistake
+				var area = state.box.width * state.box.height;
+				if (area >= 100) {
```

Comment:
I was looking at https://github.com/d3/d3-drag/blob/master/README.md#drag_clickDistance to see if we can use it it to define a click for a gage vs mouse down-drag for the draw box. Have you looked at this? This might make it easier to apply the drag interaction to the whole svg instead of layer by layer.